hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/SinkBuilder.java

```diff
@@ -0,0 +1,151 @@
+    private DistributedConsumer<S> destroyFn;
+
+    SinkBuilder() {
+    }
+
+    /**
+     * Creates and returns a custom sink that consumes items from the pipeline
+     *
+     * @param <T> the type of the data the sink will receive
+     */
+    public <T> Sink<T> build() {
+        Preconditions.checkNotNull(createFn, "createFn cannot be null");
+        Preconditions.checkNotNull(addItemFn, "addItemFn cannot be null");
+        if (flushFn == null) {
+            flushFn = noopConsumer();
+        }
+        if (destroyFn == null) {
+            destroyFn = noopConsumer();
+        }
+
+        ProcessorSupplier supplier = new Supplier<>(createFn, addItemFn, flushFn, destroyFn);
+        return new SinkImpl<>("custom-sink", preferLocalParallelismOne(supplier));
+    }
+
+    /**
+     * Sets the sink object creation function
+     *
+     * @param createSinkFn a function that takes the {@link JetInstance}
+     *                     and creates the sink. The returned sink object will be
+     *                     passed to the following {@code addItemFn}, {@code flushFn}
+     *                     and {@code destroyFn} functions.
+     */
+    public SinkBuilder<T, S> createFn(DistributedFunction<JetInstance, S> createSinkFn) {
+        this.createFn = createSinkFn;
+        return this;
+    }
+
+    /**
+     * Sets the item addition to sink function
+     *
+     * @param addToSinkFn A bi-consumer that should implement the sink logic by
+     *                    taking sink object and an item from the pipeline.
+     *                    The logic should consume the item by adding it to the sink.
+     */
+    public SinkBuilder<T, S> addItemFn(DistributedBiConsumer<S, T> addToSinkFn) {
+        this.addItemFn = addToSinkFn;
+        return this;
+    }
+
+    /**
+     * Sets the sink flush function
+     *
+     * @param flushSinkFn An optional consumer that can be used to implement
+     *                    flushing logic.
+     */
+    public SinkBuilder<T, S> flushFn(DistributedConsumer<S> flushSinkFn) {
+        this.flushFn = flushSinkFn;
+        return this;
+    }
+
+    /**
+     * Sets the sink destroy function
+     *
+     * @param destroySinkFn An optional consumer that can be used to implement
+     *                      destroy logic to clean-up resources allocated when
+     *                      creating the sink object.
+     */
+    public SinkBuilder<T, S> destroyFn(DistributedConsumer<S> destroySinkFn) {
+        this.destroyFn = destroySinkFn;
+        return this;
+    }
+
+    private static final class Supplier<S, T> extends CloseableProcessorSupplier {
+
+        private static final long serialVersionUID = -1L;
+
+        private DistributedFunction<JetInstance, S> createFn;
+        private DistributedBiConsumer<S, T> addItemFn;
+        private DistributedConsumer<S> flushFn;
+        private DistributedConsumer<S> destroyFn;
+
+        private Supplier(DistributedFunction<JetInstance, S> createFn,
```

Comment:
it would make more sense to modify WriteBufferedP to match this, rather than create yet another identical supplier